State.h

```diff
@@ -0,0 +1,82 @@
+#ifndef STATE_H
+#define STATE_H
+
+#include "BreathData.h"
+
+/*
+ *  Parent class for all instantiated state classes
+ */
+class State {
+  public:
+    virtual State* update() = 0;           // perform maintenance, and perhaps transition to new state
+    static BreathData* breath;             // holds shared information about the breath cycle
+};
+
+/*
+ *  Specific states used by the ventilator control.
+ *  Only a singleton instance of each is created, accessible via the get() method.
+ */
+
+// Standby mode.  This is the initial state.
+class OffState : public State {
+  public:
+    static State* enter();               // performs actions necessary upon entering the state
```

Comment:
By making `enter` static, you are preventing it from directly setting member variables in this class.  It feels like a weird asymmetry that `enter` belongs to the class but `update` belongs to the state.  Since every state is a singleton, they can easily both be virtual functions.

Reply:
I wanted `enter` to be static so you can call it from just the class name.  Otherwise you have to have all the singletons be globally available, right?   That's not a problem with `update` because you should already have a local pointer to it before updating.

Reply:
Sure, but there are several ways to address that:

1. Make all the singletons global.  This is not as bad as it sounds.  The state machine needs to know all of the state names anyway, and each state name is represented by a singleton, so there isn't really any loss of abstraction nor any proliferation of names that isn't already there.
2. Have a `singleton()` static function in in each class.  This is slightly cleaner, but also a bit more cumbersome, as you'd have to write `OffState::singleton().enter()` instead of just `OffState::enter()` or `offState.enter()`.
3. An interesting variation of #1 is to create a global function template, `enterState<StateClass>()`, so you would write `enterState<OffState>()` instead of `OffState::enter()`.  `enterState` would be written as:

```
template <class StateClass>
State* enterState() {
    return State::singleton().enter();
};
```
As you can see, the asymmetry really bugs me.  I think for example that the state class should hold onto all of its own state-related timers, rather than putting all of that junk into `BreathData`.  That will work best if `enter` and `update` are both virtual functions.  Note that the individual state classes can also have non-virtual functions to return, e.g., a timer value when the state was last entered or exited.

Reply:
Asymmetry resolved.